[PyTorch TB] Write raw tensor as tensor_proto

sethlee0111 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104908
* #105027

This is the first diff to support logging of raw tensors for [TensorBoard Intermediate Logging](https://www.internalfb.com/intern/wiki/TensorBoard/Intermediate_Logging/)

Ultimately, we aim to support the feature where store full tensor is stored as a tensor protobuf to TB. Protobuf contains shape, dtype, and elements of the given tensor.

1. add `tensor_proto()` to `summary.py` which takes a tensor and convert to protobuf
2. add `add_tensor()` to `writer.py`
3. formatting changes introduced by `arc lint`
-------------


Differential Revision: [D47302017](https://our.internmc.facebook.com/intern/diff/D47302017/)

sethlee0111 committed:
[PyTorch TB] Write raw tensor as tensor_proto

This is the first diff to support logging of raw tensors for [TensorBoard Intermediate Logging](https://www.internalfb.com/intern/wiki/TensorBoard/Intermediate_Logging/)

Ultimately, we aim to support the feature where store full tensor is stored as a tensor protobuf to TB. Protobuf contains shape, dtype, and elements of the given tensor.

1. add `tensor_proto()` to `summary.py` which takes a tensor and convert to protobuf
2. add `add_tensor()` to `writer.py`
3. formatting changes introduced by `arc lint`
-------------

ghstack-source-id: 194155757

Differential Revision: [D47302017](https://our.internmc.facebook.com/intern/diff/D47302017/)

[ghstack-poisoned]

vadimkantorov commented on torch/utils/tensorboard/summary.py:
```diff
@@ -303,6 +325,75 @@ def scalar(name, tensor, collections=None, new_style=False, double_precision=Fal
+            ],
+            proto_val_field: torch.view_as_real(tensor).tolist()
+            if proto_val_field == "scomplex_val" or proto_val_field == "dcomplex_val"
+            else tensor.reshape(-1).tolist(),
```
here a bit more idiomatic and modern might be `.flatten()` instead of reshape(-1)

sethlee0111 replied:
Updated, thanks!

sethlee0111 replied:
Added checks for `numel() == 1` and `numel() == 0`

sethlee0111 committed:
Update on "[PyTorch TB] Write raw tensor as tensor_proto"


This is the first diff to support logging of raw tensors for [TensorBoard Intermediate Logging](https://www.internalfb.com/intern/wiki/TensorBoard/Intermediate_Logging/)

Ultimately, we aim to support the feature where store full tensor is stored as a tensor protobuf to TB. Protobuf contains shape, dtype, and elements of the given tensor.

1. add `tensor_proto()` to `summary.py` which takes a tensor and convert to protobuf
2. add `add_tensor()` to `writer.py`
3. formatting changes introduced by `arc lint`
-------------


Differential Revision: [D47302017](https://our.internmc.facebook.com/intern/diff/D47302017/)

[ghstack-poisoned]

sethlee0111 committed:
Update on "[PyTorch TB] Write raw tensor as tensor_proto"


This is the first diff to support logging of raw tensors for [TensorBoard Intermediate Logging](https://www.internalfb.com/intern/wiki/TensorBoard/Intermediate_Logging/)

Ultimately, we aim to support the feature where store full tensor is stored as a tensor protobuf to TB. Protobuf contains shape, dtype, and elements of the given tensor.

1. add `tensor_proto()` to `summary.py` which takes a tensor and convert to protobuf
2. add `add_tensor()` to `writer.py`
3. formatting changes introduced by `arc lint`
-------------


Differential Revision: [D47302017](https://our.internmc.facebook.com/intern/diff/D47302017/)

[ghstack-poisoned]

sethlee0111 committed:
Update on "[PyTorch TB] Write raw tensor as tensor_proto"


This is the first diff to support logging of raw tensors for [TensorBoard Intermediate Logging](https://www.internalfb.com/intern/wiki/TensorBoard/Intermediate_Logging/)

Ultimately, we aim to support the feature where store full tensor is stored as a tensor protobuf to TB. Protobuf contains shape, dtype, and elements of the given tensor.

1. add `tensor_proto()` to `summary.py` which takes a tensor and convert to protobuf
2. add `add_tensor()` to `writer.py`
3. formatting changes introduced by `arc lint`
-------------


Differential Revision: [D47302017](https://our.internmc.facebook.com/intern/diff/D47302017/)

[ghstack-poisoned]

sethlee0111 committed:
Update on "[PyTorch TB] Write raw tensor as tensor_proto"


This is the first diff to support logging of raw tensors for [TensorBoard Intermediate Logging](https://www.internalfb.com/intern/wiki/TensorBoard/Intermediate_Logging/)

Ultimately, we aim to support the feature where store full tensor is stored as a tensor protobuf to TB. Protobuf contains shape, dtype, and elements of the given tensor.

1. add `tensor_proto()` to `summary.py` which takes a tensor and convert to protobuf
2. add `add_tensor()` to `writer.py`
3. formatting changes introduced by `arc lint`
-------------


Differential Revision: [D47302017](https://our.internmc.facebook.com/intern/diff/D47302017/)

[ghstack-poisoned]

sethlee0111 committed:
Update on "[PyTorch TB] Write raw tensor as tensor_proto"


This is the first diff to support logging of raw tensors for [TensorBoard Intermediate Logging](https://www.internalfb.com/intern/wiki/TensorBoard/Intermediate_Logging/)

Ultimately, we aim to support the feature where store full tensor is stored as a tensor protobuf to TB. Protobuf contains shape, dtype, and elements of the given tensor.

1. add `tensor_proto()` to `summary.py` which takes a tensor and convert to protobuf
2. add `add_tensor()` to `writer.py`
3. formatting changes introduced by `arc lint`
-------------


Differential Revision: [D47302017](https://our.internmc.facebook.com/intern/diff/D47302017/)

[ghstack-poisoned]

sethlee0111 committed:
Update on "[PyTorch TB] Write raw tensor as tensor_proto"


This is the first diff to support logging of raw tensors for [TensorBoard Intermediate Logging](https://www.internalfb.com/intern/wiki/TensorBoard/Intermediate_Logging/)

Ultimately, we aim to support the feature where store full tensor is stored as a tensor protobuf to TB. Protobuf contains shape, dtype, and elements of the given tensor.

1. add `tensor_proto()` to `summary.py` which takes a tensor and convert to protobuf
2. add `add_tensor()` to `writer.py`
3. formatting changes introduced by `arc lint`
-------------


Differential Revision: [D47302017](https://our.internmc.facebook.com/intern/diff/D47302017/)

[ghstack-poisoned]

sethlee0111 committed:
Update on "[PyTorch TB] Write raw tensor as tensor_proto"


This is the first diff to support logging of raw tensors for [TensorBoard Intermediate Logging](https://www.internalfb.com/intern/wiki/TensorBoard/Intermediate_Logging/)

Ultimately, we aim to support the feature where store full tensor is stored as a tensor protobuf to TB. Protobuf contains shape, dtype, and elements of the given tensor.

1. add `tensor_proto()` to `summary.py` which takes a tensor and convert to protobuf
2. add `add_tensor()` to `writer.py`
3. formatting changes introduced by `arc lint`
-------------


Differential Revision: [D47302017](https://our.internmc.facebook.com/intern/diff/D47302017/)

[ghstack-poisoned]

sethlee0111 committed:
Update on "[PyTorch TB] Write raw tensor as tensor_proto"


This is the first diff to support logging of raw tensors for [TensorBoard Intermediate Logging](https://www.internalfb.com/intern/wiki/TensorBoard/Intermediate_Logging/)

Ultimately, we aim to support the feature where store full tensor is stored as a tensor protobuf to TB. Protobuf contains shape, dtype, and elements of the given tensor.

1. add `tensor_proto()` to `summary.py` which takes a tensor and convert to protobuf
2. add `add_tensor()` to `writer.py`
3. formatting changes introduced by `arc lint`
-------------


Differential Revision: [D47302017](https://our.internmc.facebook.com/intern/diff/D47302017/)

[ghstack-poisoned]

sethlee0111 committed:
Update on "[PyTorch TB] Write raw tensor as tensor_proto"


This is the first diff to support logging of raw tensors for [TensorBoard Intermediate Logging](https://www.internalfb.com/intern/wiki/TensorBoard/Intermediate_Logging/)

Ultimately, we aim to support the feature where store full tensor is stored as a tensor protobuf to TB. Protobuf contains shape, dtype, and elements of the given tensor.

1. add `tensor_proto()` to `summary.py` which takes a tensor and convert to protobuf
2. add `add_tensor()` to `writer.py`
3. formatting changes introduced by `arc lint`
-------------


Differential Revision: [D47302017](https://our.internmc.facebook.com/intern/diff/D47302017/)

[ghstack-poisoned]

kunalb commented on torch/utils/tensorboard/summary.py:
```diff
@@ -323,6 +326,85 @@ def scalar(name, tensor, collections=None, new_style=False, double_precision=Fal
+        )
+
+    # type maps: torch.Tensor type -> (protobuf type, protobuf val field)
+    type_map = {
```
make this a private constant map outside the function? _TENSOR_TYPE_MAP = ... or something?

sethlee0111 replied:
Makes sense, applied changes.

kunalb commented on torch/utils/tensorboard/summary.py:
```diff
@@ -323,6 +326,85 @@ def scalar(name, tensor, collections=None, new_style=False, double_precision=Fal
+
+    if tensor.dtype in type_map:
+        proto_val_field = type_map[tensor.dtype][1]
+        proto_val_contents = (
```
For readability, I'd still this make this explicit
proto_val_contents = [] 
if proto_val_field == ...:
    proto_val_contents = ...
elif tensor.numel() == 1
    proto_val_contents  = ...

and so on

sethlee0111 replied:
I see what you mean. Made changes.

kunalb reviewed (approved):
Generally looks good to me, thanks for adding tests! some nits inline

sethlee0111 committed:
Update on "[PyTorch TB] Write raw tensor as tensor_proto"


This is the first diff to support logging of raw tensors for [TensorBoard Intermediate Logging](https://www.internalfb.com/intern/wiki/TensorBoard/Intermediate_Logging/)

Ultimately, we aim to support the feature where store full tensor is stored as a tensor protobuf to TB. Protobuf contains shape, dtype, and elements of the given tensor.

1. add `tensor_proto()` to `summary.py` which takes a tensor and convert to protobuf
2. add `add_tensor()` to `writer.py`
3. formatting changes introduced by `arc lint`
-------------


Differential Revision: [D47302017](https://our.internmc.facebook.com/intern/diff/D47302017/)

[ghstack-poisoned]

atalman commented:
@pytorchbot revert -c ignoredsignal -m "broke win-vs2019-cpu-py3 and macos-12-py3-arm64 "
